typed-racket-lib/typed-racket/infer/infer-unit.rkt

```diff
@@ -588,6 +588,24 @@
         [((Distinction: _ _ S) T)
          (cg S T obj)]
 
+        ;; XXX only barely noticed that cg takes an obj parameter
+        ;; halfway through the rebase
+        [((Con: S-pre S-post) (Con: T-pre T-post))
+         (% cset-meet (cg T-pre S-pre obj) (cg S-post T-post obj))]
+        [((FlatCon: S-pre S-post) (FlatCon: T-pre T-post))
+         (% cset-meet (cg T-pre S-pre obj) (cg S-post T-post obj))]
+        [((FlatCon: S-pre S-post) (Con: T-pre T-post))
+         (% cset-meet (cg T-pre S-pre obj) (cg S-post T-post obj))]
+        [((and (PredicateProp: (PropSet: (TypeProp: _ S-post) _))
+               (Fun: (list (Arrow: (list S-pre) _ _ _))))
+          ;; Apparently I can't just have the FlatCon case -- is the inference
+          ;; not aware of the subtyping relation?
+          (Con*: T-pre T-post))
+         (% cset-meet (cg T-pre S-pre obj) (cg S-post T-post obj))]
+        [((Fun: (list (Arrow: (list S-pre) _ _ _)))
+          (Con*: T-pre T-post))
+         (% cset-meet (cg T-pre S-pre obj) (cg S-pre T-post obj))]
+
```

Comment:
The `obj` parameter should only stay the same for recursive calls if the thing we're generating constraints for is the same.

i.e. when passing under a set-theoretic type constructor like `U`, we're still talking about the same value, so `obj` stays the same. However, if we're looking at some structural sub-compotent of the original thing (i.e. a field of a struct, the domain/range of a function, etc) then we're no longer looking at that same `obj` and so the argument should be `#f` (or just left off, since that's its default value).

I don't fully understand the `Con` types yet, but my suspicion at the moment is we're no longer generating constraints for `obj` when we're looking in the sub-components of a `Con` (though like I said, I don't "just get" them yet).

At any rate, the conservative (i.e. sound but not complete) thing to do is leave off the `obj`s in these new cases, so maybe that's best for now?